Sources/NIO/SocketChannel.swift

```diff
@@ -489,36 +500,94 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
         var rawAddressLength = socklen_t(MemoryLayout<sockaddr_storage>.size)
         var buffer = self.recvAllocator.buffer(allocator: self.allocator)
         var readResult = ReadResult.none
-
-        for i in 1...self.maxMessagesPerRead {
-            guard self.isOpen else {
-                throw ChannelError.eof
-            }
-            buffer.clear()
-
-            let result = try buffer.withMutableWritePointer {
-                try self.socket.recvfrom(pointer: $0, storage: &rawAddress, storageLen: &rawAddressLength)
+        
+        return try self.selectableEventLoop.withControlMessageBytes {
+            let controlBytes = self.reportExplicitCongestionNotifications ?
+                    $0:
+                    UnsafeMutableRawBufferPointer(start: nil, count: 0)
+
+            for i in 1...self.maxMessagesPerRead {
+                guard self.isOpen else {
+                    throw ChannelError.eof
+                }
+                buffer.clear()
+
+                var controlByteSlice = controlBytes[...]
+                var controlMessageReceiver = ControlMessageReceiver()
+                
+                let result = try buffer.withMutableWritePointer {
+                    try self.socket.recvmsg(pointer: $0, storage: &rawAddress, storageLen: &rawAddressLength,
+                                            controlBytes: &controlByteSlice,
+                                            controlMessageReceiver: { (controlMessage) in controlMessageReceiver.receiveMessage(controlMessage) })
+                }
+                switch result {
+                case .processed(let bytesRead):
+                    assert(bytesRead > 0)
+                    assert(self.isOpen)
+                    let mayGrow = recvAllocator.record(actualReadBytes: bytesRead)
+                    readPending = false
+
+                    let msg = AddressedEnvelope(remoteAddress: rawAddress.convert(),
+                                                data: buffer,
+                                                metadata: self.reportExplicitCongestionNotifications ?
+                                                    AddressedEnvelope.Metadata(ecnState: controlMessageReceiver.ecnValue) :
```

Comment:
Again, this ternary kinda sucks. Can `ControlMessageReceiver` take over the ternary for us in both cases?